reference/test/reorder/mc64_kernels.cpp

```diff
@@ -0,0 +1,406 @@
+    auto mc64_factory =
+        gko::experimental::reorder::Mc64<value_type, index_type>::build()
+            .with_strategy(
+                gko::experimental::reorder::mc64_strategy::max_diagonal_sum)
+            .on(this->ref);
+
+    auto mc64 = mc64_factory->generate(this->mtx);
+
+    auto perm_obj = this->unpack(mc64.get()).first;
+    auto perm = perm_obj->get_const_permutation();
+    ASSERT_EQ(perm[0], 1);
+    ASSERT_EQ(perm[1], 0);
+    ASSERT_EQ(perm[2], 5);
+    ASSERT_EQ(perm[3], 2);
+    ASSERT_EQ(perm[4], 4);
+    ASSERT_EQ(perm[5], 3);
+}
+
+
+TYPED_TEST(Mc64, CreatesCorrectPermutationAndScalingExampleProduct)
+{
+    using index_type = typename TestFixture::index_type;
+    using real_type = typename TestFixture::real_type;
+    using value_type = typename TestFixture::value_type;
+    using matrix_type = typename TestFixture::matrix_type;
+    using perm_type = typename TestFixture::perm_type;
+    auto mc64_factory =
+        gko::experimental::reorder::Mc64<value_type, index_type>::build()
+            .with_strategy(
+                gko::experimental::reorder::mc64_strategy::max_diagonal_product)
+            .on(this->ref);
+    auto mc64 = mc64_factory->generate(this->mtx);
+
+    auto perm = this->unpack(mc64.get()).first->get_const_permutation();
+    auto row_scaling =
+        this->unpack(mc64.get()).first->get_const_scaling_factors();
+    auto col_scaling =
+        this->unpack(mc64.get()).second->get_const_scaling_factors();
+
+    ASSERT_EQ(perm[0], 1);
+    ASSERT_EQ(perm[1], 5);
+    ASSERT_EQ(perm[2], 3);
+    ASSERT_EQ(perm[3], 4);
+    ASSERT_EQ(perm[4], 0);
+    ASSERT_EQ(perm[5], 2);
+    GKO_ASSERT_NEAR(row_scaling[0], value_type{1. / 3.}, this->tolerance);
+    GKO_ASSERT_NEAR(row_scaling[1], value_type{0.2}, this->tolerance);
+    GKO_ASSERT_NEAR(row_scaling[2], value_type{0.2}, this->tolerance);
+    GKO_ASSERT_NEAR(row_scaling[3], value_type{4. / 15.}, this->tolerance);
+    GKO_ASSERT_NEAR(row_scaling[4], value_type{0.3}, this->tolerance);
+    GKO_ASSERT_NEAR(row_scaling[5], value_type{2. / 15.}, this->tolerance);
+    GKO_ASSERT_NEAR(col_scaling[0], value_type{1.}, this->tolerance);
+    GKO_ASSERT_NEAR(col_scaling[1], value_type{1.5}, this->tolerance);
+    GKO_ASSERT_NEAR(col_scaling[2], value_type{0.9375}, this->tolerance);
+    GKO_ASSERT_NEAR(col_scaling[3], value_type{5. / 6.}, this->tolerance);
+    GKO_ASSERT_NEAR(col_scaling[4], value_type{1.}, this->tolerance);
+    GKO_ASSERT_NEAR(col_scaling[5], value_type{1.25}, this->tolerance);
+}
+
+
+TYPED_TEST(Mc64, CreatesCorrectPermutationAndScalingLargeTrivialExampleProduct)
+{
+    using index_type = typename TestFixture::index_type;
+    using real_type = typename TestFixture::real_type;
+    using value_type = typename TestFixture::value_type;
+    using matrix_type = typename TestFixture::matrix_type;
+    using perm_type = typename TestFixture::perm_type;
+    // read input data
```

Comment:
nit: I'm not sure if we usually add these comments, but just by glancing over the tests, they seem to clash with our AAA style. But removal is not necessary for me.